src/persist/src/indexed/encoding.rs

```diff
@@ -0,0 +1,564 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Data structures stored in Blobs and Buffers in serialized form.
+
+use std::cmp::Ordering;
+use std::time::SystemTime;
+
+use abomonation_derive::Abomonation;
+use differential_dataflow::trace::Description;
+use timely::progress::Antichain;
+use timely::PartialOrder;
+
+use crate::error::Error;
+use crate::storage::SeqNo;
+
+/// The structure serialized and stored as an entry in a
+/// [crate::storage::Buffer].
+#[derive(Debug, Abomonation)]
+pub struct BufferEntry {
+    /// Id of the stream this batch belongs to.
+    pub id: u64,
+    /// The updates themselves.
+    pub updates: Vec<((String, String), u64, isize)>,
+}
+
+/// The structure serialized and stored as a value in [crate::storage::Blob]
+/// storage for metadata keys.
+///
+/// TODO: Invariants
+#[derive(Clone, Debug, Abomonation)]
+pub struct BlobMeta {
+    /// The most recently assigned key name.
+    pub last_file_id: u128,
+    /// BlobFutures indexed by stream id.
+    pub futures: Vec<(u64, BlobFutureMeta)>,
```

Comment:
these are meant to have set semantics right - exactly one instance of each id and both futures and traces have the exact same sets of ids?